### PR TITLE
[feat] make core components more clearer

### PR DESCRIPTION
**Pull request by AlexandreBrg**
## Context

Internal components of the controller was stuck with an architecture making it impossible to handle telemetry from scheduler. This aims to simplify the architecture with barriers between components of the system. I didn't migrate the whole system, we still have a legacy that need to be removed.

Took me 3 hours to re-write.

## What has been done 

* Create a Core component which is used as a mediator to handle events and communicate with other internal components
* Instance handling is done through two implementation: an `InstanceService` which contains method to handle business logic for instance, an `instanceRepository` which contains methods to save data
* Core contains a reconciliation loop which can only handle a single event at a time
* Legacy events from external is converted into CoreInternalEvent
* CoreInternalEvent are meant to be used internally in the core
* Initial connection to scheduler is now using backoff to do retries
* Worker state persistency
* Add function workload definition
* Update OpenAPI with Function workload definition

Closes https://github.com/polyxia-org/project/issues/20

### CODE REVIEW

File: controller/src/core/core.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
+use crate::api::{ApiChannel, RikError, CRUD};
+use crate::core::instance::Instance;
+use crate::core::instance_repository::InstanceRepositoryImpl;
+use crate::core::instance_service::InstanceServiceImpl;
+use crate::core::worker_repository::WorkerRepositoryImpl;
+use crate::core::worker_service::WorkerServiceImpl;
+use crate::core::{InstanceService, Listener, WorkerService};
+use crate::database::RikDataBase;
+use definition::workload::WorkloadDefinition;
+use proto::common::worker_status::Status;
```
[the remark this answers is not in the document]
Comment (AlexandreBrg) [reply controller/src/core/core.rs]: Done

File: controller/src/core/instance_service.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+impl InstanceServiceImpl {
+    pub(crate) async fn new(
+        service: InstanceRepositoryImpl,
+        sender: Sender<CoreInternalEvent>,
+    ) -> Result<InstanceServiceImpl, RikError> {
+        dotenv().ok();
+        let scheduler_url = match std::env::var("SCHEDULER_URL") {
+            Ok(val) => val,
+            Err(_e) => "http://127.0.0.1:4996".to_string(),
```
Comment (hugoamalric) [inline controller/src/core/instance_service.rs:70]: Hard coded localhost ip ?
Better to put it in a variable

**Review by hugoamalric — CHANGES_REQUESTED** [review]
Could you please also resolve all warnings if possible ?

File: controller/src/core/instance_service.rs (hunk 2 above)
Comment (AlexandreBrg) [reply controller/src/core/instance_service.rs:69]: Done

**Comment by AlexandreBrg** [comment]
> Could you please also resolve all warnings if possible ?

I removed all warnings, except warnings related to dead code.

**Review by hugoamalric — APPROVED** [review]

**Review by MaloPolese — APPROVED** [review]

**Pull request merged**